### PR TITLE
fail2ban: fix 2to3 error

### PR DESCRIPTION
**Pull request by erdoukki**
Maintainer: me / @erdoukki 
Run tested: (test done on MVEBU arch53 EspressoBin & EspressoBin-ULTRA, OpenWrt 21.02.x)
Compile tested: (arch=cortexa53, host=amd64)

Issue: 2to3 support has been removed in setuptools since version 58.0.0.

Fix: openwrt#17311

Requirements: 2to3/host openwrt/packages#17429

To install/build for python3 from source, it is necessary to convert to py3
codebase before setup (invoke 2to3 or ./fail2ban-2to3 firstly).
> ./fail2ban-2to3
> python3 setup.py build

Signed-off-by: Kerma Gérald <gandalf@gk2.net>

### CODE REVIEW

**Comment by erdoukki** [comment]
@BKPepe 
Can you check, please ?

File: lang/python/host-pip-requirements/lib2to3.txt
Comment (jefferyto) [inline lang/python/host-pip-requirements/lib2to3.txt:1]: I would prefer to do something similar to Debian/Ubuntu (their `2to3` comes from a simple [script](https://salsa.debian.org/cpython-team/python3-defaults/-/blob/e429bbce303776b0357ef2636fd4b43ee0945c57/debian/2to3.py) that they maintain) rather than install a random `2to3` from PyPI.

I'll try to find time in the coming week to add a 2to3 host package, then this package can do `HOST_BUILD_DEPENDS:=2to3/host`.

Comment (jefferyto) [reply lang/python/host-pip-requirements/lib2to3.txt:1]: > …why do you say « random » ?

* Project doesn't have a readme or a public code repository, only has one maintainer and one release from 2018
* Maintainer only has 2 projects total, the other one is labelled as deprecated

Comment (erdoukki) [reply lang/python/host-pip-requirements/lib2to3.txt:1]: Great... Thanks !

Comment (erdoukki) [reply lang/python/host-pip-requirements/lib2to3.txt:1]: > If needed, I can also take a look, test, or make the proposed PR, but I don’t know when.

I try to make the package for 2to3/host...
Will propose a PR if I get success on it !

Comment (erdoukki) [reply lang/python/host-pip-requirements/lib2to3.txt:1]: Package made and compiled on master...
Waiting for your advice and review.
;-)

Comment (erdoukki) [reply lang/python/host-pip-requirements/lib2to3.txt:1]: https://github.com/openwrt/packages/pull/17429

**Comment by erdoukki** [comment]
May not compile correctly without the required éto3/host package from PR https://github.com/openwrt/packages/pull/17429

File: net/fail2ban/Makefile
Comment (jefferyto) [inline net/fail2ban/Makefile:23]:
```suggestion
HOST_BUILD_DEPENDS:=2to3/host
```

Since `2to3/host` already depends on `python3/host`.

File: net/fail2ban/Makefile
Comment (jefferyto) [inline net/fail2ban/Makefile:81]:
```suggestion
	cd $(PKG_BUILD_DIR) && ./fail2ban-2to3
```

`$(PYTHON3_PKG_SETUP_DIR)` is only needed if your package had set `PYTHON3_PKG_SETUP_DIR` earlier.

File: net/fail2ban/Makefile
Comment (erdoukki) [reply net/fail2ban/Makefile:23]: okay, agreed, thanks, modified

Comment (erdoukki) [reply net/fail2ban/Makefile:23]: But still non testable with actual master...
Will have to switch to my local 2to3 branch, because of still missing 2to3 host package.
https://github.com/openwrt/packages/pull/17429 still pending !

File: net/fail2ban/Makefile
Comment (erdoukki) [reply net/fail2ban/Makefile:81]: Okay.
Thanks for the `PYTHON3_PKG_SETUP_DIR` coorection (and for the explanation also ;-) ).
I think I have already tested something like this, with direct cd command instead of $SHELL call.
`	cd $(PKG_BUILD_DIR) && ./fail2ban-2to3`
I will test with your modification, and if it do not work, I revert with your modification like:
` 	$(SHELL) -c "cd $(PKG_BUILD_DIR) && ./fail2ban-2to3";`

File: net/fail2ban/Makefile
Comment (erdoukki) [reply net/fail2ban/Makefile]: It looks like to work, my test gets successful.
The package was built.
Working also on TestBuild...
But if the 2to3 package is missing, it might not work !
I do not understand why it is built without the 2to3 package...
I get some same issue/success with the initial package, without get the point to understand...
May be a CACHE problem or any compilation side effect !?
The issue https://github.com/openwrt/packages/issues/17311 came because of the missing 2to3 !!!
Why it might or may not work, is non-understandable for now.

File: net/fail2ban/Makefile
Comment (erdoukki) [reply net/fail2ban/Makefile]: okay with your proposal, another issue I get before, looks to be fixed now, but I still do not understand why !

**Comment by jefferyto** [comment]
Not sure why https://github.com/openwrt/packages/actions/runs/1662159088 passed even though the 2to3 package wasn't merged into master yet - ping @aparcar 

@erdoukki please rebase on current master to pick up 2to3

**Comment by erdoukki** [comment]
It now works with the add of the 196c55e93103d1e56d1336f27a5f2591b4e54fef commit from fail2ban upstream

**Comment by erdoukki** [comment]
TEST IN PROGRESS,
There is an issue on the OpenWrt target:
```
root@STARGATE:~# fail2ban-client --version
Traceback (most recent call last):
  File "/usr/bin/fail2ban-client", line 34, in <module>
    from fail2ban.client.fail2banclient import exec_command_line, sys
ModuleNotFoundError: No module named 'fail2ban'
```
I am looking at to solve it. Will update the PR with a fix.

**Comment by jefferyto** [comment]
I suggest:

1. Clone the packages repo into a new directory and check out the `openwrt-21.02` (or `openwrt-19.07`) branch
2. Copy the new package and dependencies (i.e. 2to3) into the new packages directory
3. Set up the openwrt-21.02 sdk/buildroot to use the new directory as the packages feed

File: net/fail2ban/Makefile
Comment (jefferyto) [inline net/fail2ban/Makefile:17]: Where did you see it is GPL2 only? According to the [readme](https://github.com/fail2ban/fail2ban/blob/095aeda8407b433098df35424cde2764a09566a6/README.md?plain=1#L120-L121) it is GPL2 or later.

Comment (erdoukki) [reply net/fail2ban/Makefile:17]: fixed... thanks

**Comment by erdoukki** [comment]
BUILD and TEST compatibility issue fixed with the correct version of Python.

```
root@ULTRA-5G:~# opkg install /root/fail2ban_0.11.2-4_aarch64_cortex-a53.ipk --force-reinstall
Removing package fail2ban from root...
Installing fail2ban (0.11.2-4) to root...
Configuring fail2ban.
uci: Entry not found
```

```
root@ULTRA-5G:~# fail2ban-client  --version
Fail2Ban v0.11.2
```
Tests done.
Upgrade fine.
Looks to work fine.

**Comment by erdoukki** [comment]
@jefferyto test successful on 21.02.
I will PR the cherry-picked of the commit to the 21.02.

Do we need tests and PR for 19.07 ?

**Comment by erdoukki** [comment]
@jefferyto 
All tests works ok
Build : ok
Install : ok
Upgrade : ok

**Review by jefferyto — APPROVED** [review]
Thanks!

**Pull request merged**
